### DISCUSSION TITLE
[QUESTION] Need some clarification regarding tensor parallelism on Multimodal example

### DISCUSSION THREAD

**Discussion by luisfrentzen** [NVIDIA/Megatron-LM#1197] [Q&A]
Hi I am new and trying to learn how to use the framework. I am currently following the multimodal guide in the examples section. I have some questions that I need clarification on. I hope it is okay and I am sorry if the answer should've been obvious.

1. While converting checkpoint to megatron format with the guide on this page [docs/llama_mistral.md](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/llama_mistral.md), it says that the TP parameter must be set correctly. For example, Llama-3-70B must have TP = 8. Where does that number come from? What if I want to use TP less/more than 8 for the model Llama-3-70B? Would that work?
![image](https://github.com/user-attachments/assets/db1c7523-b1a7-4508-8c99-2533adb208e8)

2. Does having target TP = 4 mean that the model weight must be shared between no more and no less than 4 GPUs?
3. What is the difference between tensor parallelism and pipeline parallelism argument?

It would be great if someone could provide a bit of explanation. Please mention if you think there is any resource that would answer these questions as well. Thank you very much.

**Comment by Advait251206** [comment]
These are good questions—Tensor Parallelism (TP) and Pipeline Parallelism (PP) can be confusing when you're first getting started with Megatron-LM.

### 1. Why does the guide say Llama-3-70B must use TP = 8?

That recommendation is primarily for the **checkpoint conversion example**, not because the model is fundamentally restricted to TP = 8.

The value comes from how the pretrained checkpoint was originally partitioned (or from the recommended configuration used by NVIDIA). During conversion, the script needs to know how tensor-sharded weights should be arranged.

In general:

- **TP = 1** means each tensor is stored as a whole.
- **TP = N** means supported weight matrices are split across `N` GPUs (or checkpoint shards).

If you're converting from a Hugging Face checkpoint (which is usually unsharded), the converter can often generate checkpoints for different TP sizes, provided the model architecture supports it and the hidden dimensions are divisible by the requested TP size.

For example, if the hidden size is divisible by 4 and 8, both TP=4 and TP=8 are generally valid choices. The limitation is usually the tensor dimensions, not the model itself.

### 2. Can I use TP smaller or larger than 8 for Llama-3-70B?

Potentially, yes.

The main requirements are:

- the model dimensions must be divisible by the TP size,
- the Megatron implementation you're using must support that configuration,
- and you must launch training/inference with the same TP size that the checkpoint expects (or reconvert the checkpoint).

You generally **cannot** load a TP=8 checkpoint while running with TP=4 unless you first convert or reshard the checkpoint.

### 3. Does TP=4 mean the model must run on exactly 4 GPUs?

Not necessarily.

TP defines the size of **one tensor-parallel group**, not the total number of GPUs.

For example:

- 4 GPUs with TP=4 → one TP group
- 8 GPUs with TP=4 → two TP groups (assuming no pipeline parallelism)
- 16 GPUs with TP=4 and PP=2 → two pipeline stages, each using TP groups of four GPUs

So TP determines how tensors are partitioned within a model replica.

### 4. What's the difference between Tensor Parallelism and Pipeline Parallelism?

They're parallelizing different things.

**Tensor Parallelism (TP)** splits **individual layers** across multiple GPUs.

For example, if a large linear layer is too large for one GPU, TP divides its weight matrix across multiple GPUs, and each GPU computes part of the operation before combining the results.

**Pipeline Parallelism (PP)** splits the **layers of the network** across GPUs.

For example:

- GPU 0: Layers 1–20
- GPU 1: Layers 21–40
- GPU 2: Layers 41–60
- GPU 3: Layers 61–80

During training, microbatches flow through these pipeline stages.

So, in short:

- **TP:** split *within* a layer.
- **PP:** split *between* layers.

These techniques are often combined for very large models.

### Resources

The following are good references if you want a deeper understanding:

- The Megatron-LM paper describing tensor and pipeline parallelism.
- NVIDIA's Megatron-Core documentation, especially the sections on model parallelism.
- The original GPipe and PipeDream papers for pipeline parallelism concepts.

Once the concepts click, the checkpoint conversion options make much more sense because they're simply determining how model weights are partitioned to match the intended parallel configuration.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
